### DISCUSSION TITLE
TypeError [ERR_INVALID_STATE]: Invalid state: ReadableStream is already closed

### DISCUSSION THREAD

**Discussion by msoyka** [vercel/next.js#55027] [App Router]
I am quickly throwing together a POC using Next.js new app router, and am making sequential api request in the layout file, and am getting the error `TypeError [ERR_INVALID_STATE]: Invalid state: ReadableStream is already closed`. Is this something to do with next.js cacheing? Curious if anyone has seen this error before and where I might start my efforts to fix.

Code Snippet:
```
export default async function RootLayout({children}: {
  children: React.ReactNode;
}) {
  const session = await getServerSession(nextAuthOptions);

  const userActiveCart = await getUserActiveCart({
    customerId: Number(session?.customerId),
    accessToken: session?.accessToken || "",
  });

  const featureFlags = await getFeatureFlags({
    accessToken: session?.accessToken || "",
    storeId: userActiveCart?.storeId || "",
  });

  return (
    <html lang="en">
      <body className={inter.className}>
            <pre>
              {JSON.stringify(userActiveCart, null, 2)}
              <br />
              <br />
              {JSON.stringify(session, null, 2)}
              <br />
              <br />
              {JSON.stringify(featureFlags, null, 2)}
            </pre>
      </body>
    </html>
  );
}
```

**Comment by devjiwonchoi** [comment]
Assuming that `{children,}` is a typo, the overall code seems alright.

Can you provide a repro?

**Reply by msoyka** [reply]
That is correct. Simplified the layout before posting it here. Removed it for clarity.

**Comment by icyJoseph** [comment]
What Node.js version have you got?

This looks like a stale stream object that's being reused. Like for example, if you'd do:

```js
const res = await fetch(/* something */)
cache.add('foo', res)
```

And then called `res.json()` twice, thought that'd yield another error.

It is not really easy to see why this would happen on your snippet though.

**Reply by msoyka** [reply]
I am using node `v18.16.0`, but each res.json() is encapsulated within its own function so don't think that is happening, but is always nice to know that calling `.json` twice would create this error. Will verify that this is indeed not happening.

**Reply by icyJoseph** [reply]
It'd be a different error message AFAIK, but sounds like a somewhat similar issue

**Comment by statusunknown418** [comment]
for me it just happens if I'm on the dev server and reload the page really fast or do a hard reload (`cmd` + `R`)

**Reply by msoyka** [reply]
This is interesting. I am doing the hard reload a lot at the moment as I am testing the caching functionality of the requests as well as my sessions that are being set with Next-Auth. Glad, not glad, that this is also happening to others from just basic dev server reloads. Will have to verify with a production build I imagine.

**Reply by statusunknown418** [reply]
yea, it kinda seems weird to have this errors anyway

**Comment by msoyka** [comment]
Trying to get together a simplified repo together in hopes it might help give better context to answering the question. Appreciate all the quick responses! 👍

**Comment by asp3** [comment]
having the same issue on production build logs, not sure if its causing our crashes though.

**Reply by statusunknown418** [reply]
I think it's not the reason for app crashes but more like warnings

**Comment by msoyka** [comment]
I can now reproduce it everytime while running next.js app with `dev`.

**Reproduce Error:**
1. Add Suspense to `RootLayout`.
2. Run request from `GlobalHeader` server-component
3. Get Error after next.js message `Fast Refresh had to perform a full reload.`.

**Ways To Stop The Error:**
- Removing the Suspense within the root layout, but keeping the request in `GlobalHeader`
- Keep Suspense, but remove request from `GlobalHeader`



```
// `layout.tsx`

import type { Metadata } from "next";
import { Inter } from "next/font/google";
import { getServerSession } from "next-auth/next";
import { Suspense } from "react";

import StyledComponentsRegistry from "@src/lib/StyledComponentsRegistry";
import { nextAuthOptions } from "@api/auth/[...nextauth]/options";
import NextAuthHandler from "@src/lib/next-auth/NextAuthHandler";
import { GlobalHeader } from "@src/components/GlobalHeader";

import Loading from "./loading";

const inter = Inter({ subsets: ["latin"] });

export const metadata: Metadata = {
  title: "Next.js App",
  description: "App Created with next.js",
};

export default async function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  const session = await getServerSession(nextAuthOptions);

  return (
    <html lang="en">
      <body className={inter.className}>
        <main>
          <Suspense fallback={<Loading />}>
            <NextAuthHandler session={session} />
            <StyledComponentsRegistry>
              <GlobalHeader />
              {children}
            </StyledComponentsRegistry>
          </Suspense>
        </main>
      </body>
    </html>
  );
}
```


```
// Global Header Component `GlobalHeader.tsx`

import "server-only";

import { getServerSession } from "next-auth/next";
import { nextAuthOptions } from "@api/auth/[...nextauth]/options";
import { getUserActiveCart } from "@src/lib/services/getUserActiveCart";
import { Login } from "@src/components/Login";

// import styles from "./GlobalHeader.module.css";

export const GlobalHeader = async () => {
  const session = await getServerSession(nextAuthOptions);

// When i run this request within Suspense, the error is thrown
  const userActiveCart = await getUserActiveCart({
    customerId: Number(session?.customerId),
    accessToken: session?.accessToken || "",
  });

// When i use static data, rather than the userActiveCart request above, I dont get the error.
  // const userActiveCart = {
  //   storeId: null,
  //   activeCartId: null,
  // };

  return (
    <div>
      <Login session={session} />
      <div>
        <ul>
          <li> Store ID: {userActiveCart?.storeId || "n/a"}</li>
          <li>Active Cart ID: {userActiveCart?.activeCartId || "n/a"}</li>
        </ul>
      </div>
      <div>
        <ul>
          <li>User: {session?.user?.name || "Guest"}</li>
          <li>Customer ID: {session?.customerId || "N/A"}</li>
          <li>Customer UUID: {session?.customerUuid || "N/A"}</li>
          <li>Email: {session?.user?.email || "N/A"}</li>
          <li>isAuthenticated: {`${!!session}`}</li>
        </ul>
      </div>
    </div>
  );
};

```

```
// Data Request

import "server-only";

interface GetUserDetailsProps {
  customerId: number;
  accessToken: string;
}

interface UserActiveCart {
  activeCartId: string | null;
  storeId: string | null;
}

function wait(milliseconds: number) {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
}

export async function getUserActiveCart({
  customerId,
  accessToken,
}: GetUserDetailsProps) {
  try {
    const result = await fetch("[graphqlEndpoint]", {
      method: "POST",
      cache: "no-cache",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${accessToken}`,
      },
      body: JSON.stringify({
        query: `[query string]`,
        variables: {
          customerId,
        },
      }),
    });

    const { data } = await result.json();
    
    await wait(5000);

    const cart = data?.carts?.[0];
    return {
      activeCartId: cart?.cartId || null,
      storeId: cart?.storeId || null,
    } as UserActiveCart;
  } catch (error) {
    console.log("Error getUserActiveCart: ", error);
  }
}

```

**Reply by msoyka** [reply]
Getting an additional error as well.

```
- error Error: The render was aborted by the server without a reason.
- error TypeError [ERR_INVALID_STATE]: Invalid state: ReadableStream is already closed
```

**Reply by msoyka** [reply]
Another thing to callout is the `await wait(5000)` in the fetch. Trying to simulate longer requests to validate suspense fallbacks.

**Reply by msoyka** [reply]
Also seems to happen when i reload to quickly. Even in production build.

**Reply by sui-kee** [reply]
thanks. it work for me

**Comment by steamwings** [comment]
I only see this when on next `13.4.19`. I tried downgrading to `13.4.10` and the error does not appear. (Re-upgraded to  `13.4.19` to double check and it appeared again.) Haven't tried other minor versions.

**Reply by msoyka** [reply]
Same for me. Downgrading to `13.4.10` removed the error. Does seem like the logging changed from that time since I am now seeing detailed request caching info, so curious if there error is still there, just being hidden from the console.

```
-  ┌ GET / 200 in 5681ms
      │
        ├──── GET https://dev../auth/realms/customer/pr.. 200 in 123ms (cache: MISS)
            │
             └── 1 level ── POST https://api.dev.cloud/graphql 200 in 179ms (cache: MISS)
```

**Reply by alexvobo** [reply]
Running into the same issue (app router + next-auth). Downgrading removed the error for me

**Reply by thexpand** [reply]
I'm running into the same issue. I'm not using `next-auth`, but using `@auth0/nextjs-auth0`. I don't know if it's related to auth at all.

**Reply by tem-ctrl** [reply]
Same error here whereas not using any special package. I think this issue is a bug in the latest version on NextJs. I'm using v. 4.13.19.

**Reply by j3rryl** [reply]
I was using version 13.4.19. Downgrading to 13.4.10 seems to have removed it as well. Maybe it's the error being hidden from the logs. But getting same results as @msoyka . Any ideas to the cause, or potential fix? 
-  ┌ GET /dashboard/movies?_rsc=a3ulp 200 in 1289ms
   │
   └──── GET http://127.0.0.1:3000/api/movies 200 in 236ms (cache: MISS)


-  ┌ GET /dashboard/movies?_rsc=a3ulp 200 in 1362ms
   │
   └──── GET http://127.0.0.1:3000/api/movies 200 in 299ms (cache: MISS)

**Comment by dstarosta** [comment]
Same issue. Reverting to 13.4.12 fixes the problem.

**Reply by adarsh-drishya** [reply]
i am worried if its worth to move this to prod right now

**Comment by toonverbeek** [comment]
Same issue here. I'm not using any `<Suspense />` boundaries in my app, but I _am_ doing multiple fetches in the components where this is happening. Can confirm that downgrading removes the error.

**Comment by douglasrcjames** [comment]
Running into this same issue  on "next": "^13.4.19", but not worth downgrading since I do not see any issues with my app.

I am also using "next-auth": "^4.23.1", which seems a few others are as well, but probably not directly related.

**Comment by vedantroy** [comment]
I'm getting this error as well, and it is crashing my production server:
```
TypeError [ERR_INVALID_STATE]: Invalid state: ReadableStream is already closed
    at new NodeError (node:internal/errors:405:5)
    at ReadableByteStreamController.enqueue (node:internal/webstreams/readablestream:1152:13)
    at n (/app/.next/server/chunks/892.js:38426:141)
    at R (/app/.next/server/chunks/892.js:39206:90)
    at mb (/app/.next/server/chunks/892.js:39195:35)
    at Timeout._onTimeout (/app/.next/server/chunks/892.js:39001:16)
    at listOnTimeout (node:internal/timers:569:17)
    at process.processTimers (node:internal/timers:512:7) {
  code: 'ERR_INVALID_STATE'
}
TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11576:11)
    at async invokeRequest (/app/node_modules/next/dist/server/lib/server-ipc/invoke-request.js:17:12)
    at async invokeRender (/app/node_modules/next/dist/server/lib/router-server.js:254:29)
    at async handleRequest (/app/node_modules/next/dist/server/lib/router-server.js:447:24)
    at async requestHandler (/app/node_modules/next/dist/server/lib/router-server.js:464:13)
    at async Server.<anonymous> (/app/node_modules/next/dist/server/lib/start-server.js:117:13) {
  cause: SocketError: other side closed
      at Socket.onSocketEnd (/app/node_modules/next/dist/compiled/undici/index.js:1:63301)
      at Socket.emit (node:events:526:35)
      at endReadableNT (node:internal/streams/readable:1359:12)
      at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
    code: 'UND_ERR_SOCKET',
    socket: {
      localAddress: '127.0.0.1',
      localPort: 37256,
      remoteAddress: '127.0.0.1',
      remotePort: 36341,
      remoteFamily: 'IPv4',
      timeout: undefined,
      bytesWritten: 2315,
      bytesRead: 0
    }
  }
}
TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11576:11)
    at async invokeRequest (/app/node_modules/next/dist/server/lib/server-ipc/invoke-request.js:17:12)
    at async handleRoute (/app/node_modules/next/dist/server/lib/router-utils/resolve-routes.js:283:45)
    at async resolveRoutes (/app/node_modules/next/dist/server/lib/router-utils/resolve-routes.js:483:28)
    at async handleRequest (/app/node_modules/next/dist/server/lib/router-server.js:320:102)
    at async requestHandler (/app/node_modules/next/dist/server/lib/router-server.js:464:13)
    at async Server.<anonymous> (/app/node_modules/next/dist/server/lib/start-server.js:117:13) {
  cause: Error: read ECONNRESET
      at TCP.onStreamRead (node:internal/stream_base_commons:217:20) {
    errno: -104,
    code: 'ECONNRESET',
    syscall: 'read'
  }
}
```

**Reply by dstarosta** [reply]
It looks like newer versions of Next.js are using "undici" instead of "Node fetch" which is what seems to be causing the problem.

**Reply by icyJoseph** [reply]
@dstarosta that has been the case since last year. Rather, Node.js based its so-called native fetch on undici

**Reply by icyJoseph** [reply]
@vedantroy that looks like the custom protocol used by Next.js to talk to workers. It typically uses a random port on your system to do that. 

It might just be that your system doesn't allow next.js to open ports. Or that something is failing on that communication channel.

https://github.com/vercel/next.js/discussions/51691#discussioncomment-6341657

**Comment by icyJoseph** [comment]
Could people please answer these questions:

- Node.js and Next.js versions
- Repository that fails?

**Comment by willianrod** [comment]
I'm having the same errors and I'm not using `next-auth`:

Node: v18.16.0
Next: v13.4.19
OS: MacOS Ventura 13.4.1 (c)
Chip: Apple M2 Pro

```
error TypeError [ERR_INVALID_STATE]: Invalid state: ReadableStream is already closed
```

```
TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11457:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async invokeRequest (/path/to/app/node_modules/next/dist/server/lib/server-ipc/invoke-request.js:17:12)
    at async invokeRender (/path/to/app/node_modules/next/dist/server/lib/router-server.js:254:29)
    at async handleRequest (/path/to/app/node_modules/next/dist/server/lib/router-server.js:447:24)
    at async requestHandler (/path/to/app/node_modules/next/dist/server/lib/router-server.js:464:13)
    at async Server.<anonymous> (/path/to/app/node_modules/next/dist/server/lib/start-server.js:117:13) {
  cause: Error: connect ECONNREFUSED ::1:62069
      at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1494:16) {
    errno: -61,
    code: 'ECONNREFUSED',
    syscall: 'connect',
    address: '::1',
    port: 62069
  }
}
```

**Reply by icyJoseph** [reply]
Mmm I wonder if it has to do with the IPv6 format... or if your system is somehow not allowing Node.js to open ports beyond a certain range?

**Comment by manovotny** [comment]
We just released [Next.js 13.5](https://nextjs.org/blog/next-13-5), which contains [over 438 bugs patched since 13.4](https://nextjs.org/blog/next-13-5#other-improvements).

Can anyone please ensure you're using 13.5.x and try again?

And if you're still experiencing issues, please answer the following:

* Node.js version
* Next.js version
* OS version
* Runtime version (Edge or Node.js)
* Link to a repository that has a minimal failing reproduction

Thank you! 🙏

**Reply by willianrod** [reply]
I updated to the latest version (also tested canary) and the issue still happening.

Also worth commenting, it seems that the `TypeError: fetch failed` issue is fixed.

I created an issue to track the issue. https://github.com/vercel/next.js/issues/55608

**Reply by manovotny** [reply]
@willianrod thank you! This is super helpful. I was able to reproduce on my end. Let me fire off some internal conversations.

**Reply by Cyral** [reply]
13.5 is actually the first time I have seen this error

**Reply by Inzendis** [reply]
Im on 13.5.1 and started getting this. I haven't done anything crazy. Just nesting layouts where I get ServerSession from Next Auth in some layouts that are nested. Before doing that nesting and upgrading to the newest Next version today I wasn't getting this error.

The WebAPP PWA is working as intended though I can quickly reproduce the error if I CTRL + S on any file. If I refresh the browser it doesn't reproduce the error no matter how much I spam it.

So Im guessing this is directly server side.

**Reply by Jonnotie** [reply]
Installed 13.5.3 and it's still happening here

**Reply by msalahz** [reply]
Same here, I can reproduce the issue on v13.5.3

**Reply by Samueljh1** [reply]
Same issue here on v13.5.3

**Reply by msalahz** [reply]
Can't see it anymore on v13.5.4

**Reply by sunaryohadi** [reply]
seem re-appear again in v.13.5.5

**Comment by parabuzzle** [comment]
I am seeing this same error that @willianrod mentioned quite a bit using yarn dev with turbo repo.. I have seen this happen immediately after getting a memory exhaustion error. and when I run `free` I have very little memory left. I can't hit any of the running application in my turbo repo after this happens without stopping and restarting the dev stack. 

The next time it happens I'll get the data, but it seems to be a leak somewhere in the server.

**Comment by kitchenbeats** [comment]
Im getting the same error, from v13.4.19 and up to the latest release. As mentioned, it happens on a reload.

**Comment by ccxdev** [comment]
Should be fixed on React side

Related: https://github.com/vercel/next.js/issues/55608

**Comment by nnennajohn** [comment]
Same issue here. Using 
Next: v13.5.5-canary.8
Node: v18.17.1
OS: MacOS Ventura 13.4.1 (c)
Chip: Apple M2 Pro

Using supabase auth:

<img width="686" alt="Screenshot 2023-10-12 at 9 33 02 PM" src="https://github.com/vercel/next.js/assets/1761197/3dd2bb95-7908-45d2-81c1-19ae55f2ab91">
<img width="838" alt="Screenshot 2023-10-12 at 9 33 18 PM" src="https://github.com/vercel/next.js/assets/1761197/3ddcdada-cd83-496e-91a4-ec3feddcadab">

**Comment by rabin999** [comment]
Any Update On this ?

**Comment by Driesigner** [comment]
I was getting this error with next 14.1.3 using app-router. I was getting this error when using the useTranslations hook from next-intl in an async component. Replacing the hook by the server-side await getTranslations function fixed the issue.

Before solving (error occurs):
```typescript
export const ParagraphCaseStudy: FC<Props> = async ({ paragraph }) => {
  const t = useTranslations();
}
```

Solution:
```typescript
export const ParagraphCaseStudy: FC<Props> = async ({ paragraph }) => {
  const t = await getTranslations();
}
```

Not sure if this helps, but it seems it might have something to do with doing client stuff on server-components?

**Reply by ScreamZ** [reply]
Damn the log is so bad about it! I'm facing it a lot of time and forget about it.

**Reply by amannn** [reply]
Thanks for sharing your solution @Driesigner!

Note that the React ESLint plugin will warn for using hooks in async components in the future, hopefully that should help people to avoid running into this (see https://github.com/vercel/next.js/discussions/55027#discussioncomment-10018945).

**Reply by kisstamasj** [reply]
i can agree with you, i had the same problem, and it caused by the next-intl library, when i used useTranslations hook inside an async server component.

**Reply by clemenshimmer** [reply]
Had the exact same issue with `getLocale` and `useLocale` in a server component.

**Comment by maslowivan** [comment]
14.2.4 still happening

**Reply by awbranch** [reply]
14.2.4 still happening

**Reply by PutawanDE** [reply]
It's still happening to me too on production. It was reported by Sentry. Should we open an issue?

**Reply by ayhan02** [reply]
When I switched to 14.3.0-canary.59, I no longer had the error

**Comment by ncostaramos** [comment]
If it helps anyone, in usecase, after spending 3 days debugging, found a not-found.tsx file calling out next-intl useLocale() function.

So, I was using useLocale() instead of the await getLocale() for a server component.

This caused a suspended thenable in dev and, strangely, when building the application no errors were found. After built and running, a 500 internal error appeared with "Invalid state: ReadableStream is already closed".

The issue was an error in code but the debugging was quite painful.

**Reply by ScreamZ** [reply]
Already reported to the owner : https://github.com/amannn/next-intl/discussions/1150 @amannn see people lose much time ;'(

**Reply by amannn** [reply]
@ScreamZ I've checked again, and `eslint-plugin-react-hooks` contains the warning to [disallow hooks in async functions](https://github.com/facebook/react/pull/27045) in the latest release candidate versions:

![Screenshot 2024-07-11 at 11 16 15](https://github.com/vercel/next.js/assets/4038316/9c928438-5ec6-4634-a22c-7e7f49ab32f0)

Hopefully the React 19 RC progresses in a timely manner for people to stop bumping into this!

**Comment by Christophvh** [comment]
Same Error on "14.2.13"

**Comment by ollebergkvist** [comment]
I can confirm that this is still happening in: 15.0.1 as well
